### PR TITLE
feat(content): anchoring primitive — durable pointers into prose (Content Tags phase 0)

### PR DESCRIPTION
**Pull request by 2witstudios**
## What

Adds `packages/lib/src/content/anchoring/` — a page-type-agnostic primitive for keeping a durable pointer into prose that users then edit. Five pure modules and their tests. **Nothing else**: no schema, no service, no routes, no UI, and deliberately not yet wired into the save path.

## Why

Search in PageSpace is `ILIKE '%word%'` over `pages.title`/`pages.content` with relevance scored in JavaScript (`apps/web/src/app/api/search/route.ts`). There is no FTS infrastructure — no tsvector, no pg_trgm, no GIN index — and no RAG infrastructure — no chunks, no embeddings, no vector column — anywhere in the repo. AI retrieval is the model calling those same ILIKE/regex tools.

Content tagging is the first step out of that: a declared signal search and future retrieval can filter and rank on, independent of how good the lexical matcher is.

Tags have to attach to content *inside* a page, not just to pages. That creates the anchoring problem, and it is the *same* problem per-content comments will have later — so anchoring lands as its own primitive that tags are merely the first consumer of.

This is **Phase 0** of the Content Tags epic (`etvdzkdkq4rtzw7hq1t99vix`, drive `omziyxp4skckh7ixi2sxzhuk`).

## How — two mechanisms, in priority order

**PRIMARY: `portAnchor(anchor, oldText, newText)` forward-ports offsets through the diff.** This is GitHub's model and it is the exact one. `applyPageMutation` (`apps/web/src/services/api/page-mutation-service.ts:66`) is a single chokepoint holding `previousContent` and `nextContent` in the same transaction, and it writes a `page_versions` row on *every* mutation, so the revision chain is complete rather than sparse. Given both sides of a change the mapping is exact — no searching, no heuristics.

Crucially it is unambiguous when the quoted text occurs more than once, which a text search cannot resolve at all. `reanchor.test.ts` proves this directly: two copies of the same quote with byte-identical 32-char context on both sides, and a 200-char insertion between them. `portAnchor` lands on the right copy; `resolveAnchor`, given only the new text, lands on the **wrong** one and can only report that it was ambiguous. Both halves are asserted, so the test fails if the diff path regresses *or* if the fixture stops being genuinely ambiguous.

**FALLBACK: `resolveAnchor(text, anchor)` repairs from the quote plus context.** Used where there is no predecessor to diff against: `convert-content-mode` rewrites the whole blob through Turndown/marked, restore-from-version jumps revisions, and imports and backfills arrive with nothing behind them. Three strategies, first hit wins — position, then quote+context (nearest-to-hint when several match), then a bounded fuzzy match; below the similarity floor the anchor is `orphaned`.

### Both paths are deterministic *and* bounded

These pull in opposite directions, and the PR went through both failure modes before settling. `diff-match-patch` defaults to a one-second deadline, and an expired deadline does not fail — it returns a valid but **suboptimal** diff, so the same edit segments differently on a fast machine than on a loaded one and confidence wobbles across the orphan threshold. Removing the deadline fixes that and introduces an unbounded quadratic diff: measured here, two fully dissimilar strings take 142ms at 2KB, 558ms at 4KB and **13.9s at 20KB**, and `portAnchor` diffs once *per anchor*.

So neither bound is on the clock. Both are on the **input**, which is deterministic and bounded at once:

- `portAnchor` strips the common prefix and suffix itself (linear, `changedRegionSize`) and degrades to the repair path above `MAX_EXACT_DIFF_REGION` (2048 chars of actual change). A real save never approaches it — a 60KB document with a 14-character insertion measures 0ms.
- The fuzzy stage declines quotes over 1024 chars. Strategies 1 and 2 are linear and still run at any size.

A contract test asserts both deadline pins and confines the `diff-match-patch` import to the single module that pins them, since the real risk is the *next* module constructing its own instance.

### Other design notes

- **Why not just GitHub's "outdated" flag.** Version blobs are pruned after 30 days by `apps/web/src/app/api/cron/retention-cleanup/route.ts`, so "frozen against the original snapshot" expires. And a tag is not a review comment: a PR comment's job ends at merge so decay is free, whereas a tag has to keep working for retrieval indefinitely. Outdating on every touching edit would make actively-edited pages shed tags while stale pages keep theirs — backwards for RAG. So anchors orphan only when the text is genuinely destroyed.
- **No block IDs.** TipTap v3 serialises to an HTML string, not ProseMirror JSON (`RichEditor.tsx:46`); the `UniqueID` extension isn't loaded; there's no Yjs/CRDT anywhere; and `convert-content-mode` destroys HTML attributes. Attribute-based anchors are not durable here.
- **The HTML tokenizer is hand-rolled on purpose.** Offsets are measured against a *projection*, never the stored blob, and that projection must be byte-identical on server and client. `DOMParser` in the browser and something else in Node would disagree about whitespace, entities and implied elements — and every anchor in the product would silently drift. One implementation, shared.
- **No new dependency.** `diff-match-patch` was already here, wrapped by `packages/lib/src/content/diff-utils.ts`, whose `DiffChange` already carries `originalStart/originalEnd/newStart/newEnd` — that *is* the forward-port map.

## Files

| Module | Role |
|---|---|
| `types.ts` | `TextAnchor`, `SheetCellAnchor`, `AnchorResolution` |
| `text-projection.ts` | `projectContent(content, contentMode)` — the deterministic flattening offsets index into |
| `anchor.ts` | `createAnchor(text, start, end, opts)` |
| `reanchor.ts` | `portAnchor(anchor, oldText, newText)` — primary |
| `resolve.ts` | `resolveAnchor(text, anchor)` — fallback |

Every function is pure: no db, no fetch, no clock, no randomness, no DOM, no env. Asserted by `__tests__/purity.test.ts`, which mirrors the contract in `page-webhook-core.ts` and additionally pins the only non-relative import in the entire core to `diff-match-patch`.

## Validation

```
bun run typecheck        Tasks: 17 successful, 17 total     # monorepo-wide, CI's gate
bun run lint             Tasks: 15 successful, 15 total
bun run knip:check       [ok] knip: 4 issue(s), all within baseline (4)
vitest src/content/anchoring    Test Files 5 passed | Tests 89 passed
```

CI on the head commit: **Unit Tests, Lint & TypeScript Check, and E2E all green.**

Full `@pagespace/lib` suite locally: `9336 passed, 9 failed, 112 skipped`. Every failure is a `*.integration.test.ts` needing a live Postgres (auth handoff, billing, gdpr-eraser, page-viewers, jti-revocation, …) — pre-existing and environment-only; none in `content/`, and CI's Unit Tests job is green. Coverage thresholds hold (lines 85 / branches 94 / functions 88 / statements 85); the new modules score **100% statements, 100% functions, 98.72% branches**.

### Mutation checks — 43 mechanisms, 42 verified red

Coverage percentages don't prove a mechanism is guarded, so each was broken in source, the suite run, and then restored (baseline re-verified PASS, `git diff` clean afterwards).

Six were caught *only* this way, having survived a green suite: the `offsetWithin` term in the nearest-occurrence tie-break (a bug I introduced while refactoring and would otherwise have shipped), the common-prefix half of the changed-region measurement, the negative-start rejection, the two safe-integer offset checks — each masked by its sibling until tested independently — and a whitespace fixture that only caught its bug once padded with adjacent spaces.

**One mechanism is not verified and I'm not claiming it is:** anchoring the whitespace regex is defence in depth behind the entity-Map fix, and with the Map in place I could not construct an input that reaches `isWhitespace` with more than one character. It stays because a predicate named for a character should test a character, not a substring.

named-entity Map (no prototype chain) · textSimilarity input bound · caret-exact requires confidence 1 · tiptap/HTML whitespace sets agree · forward-port offset mapping · orphan-when-nothing-survived · drifted-hint repair before porting · repaired-never-claims-exact · caret point mapping · position check · negative-start rejection · end-past-length rejection · inverted-range rejection · safe-integer rejection (start, end) · nearest-occurrence tie-break · the tie-break's `offsetWithin` term · prefix+exact+suffix context search · caret context corroboration · caret repair from prefix+suffix · fuzzy similarity floor · fuzzy quote-length cap · 32-bit probe truncation · changed-region cap · common-prefix strip · common-suffix strip · both diff-deadline pins · block-tag newlines · script/style dropping · raw-text tag-name boundary · case-insensitive raw-text scan · entity decoding · attribute quoting only after `=` · whitespace collapsing · literal NBSP as whitespace · surrogate rejection · NUL rejection · `contentMode` narrowing · bounded context capture · offset clamping · `hashText` character mixing.

### What review turned up

Worth reading if you only read one section. Several of these were defects a green suite, full coverage, and my own re-reading had all missed:

- **Both diff paths were machine-speed dependent** — see above.
- **Removing the deadline made the diff unbounded and quadratic** — see above.
- **An apostrophe in an unquoted attribute value** opened a quote state that never closed, so one malformed tag (`<div data-x=it's fine>`) projected the entire page to the empty string and orphaned every anchor on it.
- **A caret anchor claimed `exact` with confidence 1 against unrelated content**, because `slice(start, end) === ''` holds at every in-range offset. Corroborating against the context fixed that and, as a bonus, made carets genuinely repairable — prefix and suffix are adjacent precisely when the quote between them is empty.
- **`</scripture>` ended a `<script>` skip early**, projecting the rest of the script body as visible prose (verified: `a\nleaked\nb` before the fix).
- **A literal U+00A0 and `&nbsp;` projected differently**, so a `convert-content-mode` round trip could shift every offset after the first one it touched — on precisely the path that has no diff to forward-port through.
- **NUL and lone surrogates were decoded**, producing projections Postgres cannot store in the jsonb column an anchor's `exact` is headed for.
- **Offsets from storage were trusted.** `String.slice` forgives fractional, non-finite, negative and inverted ranges by coercing or truncating, and handed the bad values back as positions.
- **`&__proto__;` was decoded.** `NAMED_ENTITIES` was an object literal, so the lookup reached the object prototype, and that name is short enough to pass the entity length guard. `Object.prototype` isn't `undefined`, so it was accepted as a decoded value — then, because it stringifies to `[object Object]` and the whitespace regex was unanchored, it was classified as whitespace and silently **swallowed**. Worse than visible garbage: it put every subsequent offset out by eleven characters against a projection that still looked entirely reasonable. `Readonly<Record<string, string>>` actively concealed it from the compiler.
- **`textSimilarity`'s bound lived at its call site**, but the module is a package export, so an outside caller got the unbounded quadratic version. A bound only some callers observe is not a bound.
- **A repaired caret could claim `exact` with confidence 0.95**, contradicting `CONFIDENCE.position` where `exact` always means 1 — so a consumer could not gate on either field alone.
- **The purity test's own import scan** only matched `import … from`, so a side-effect import, re-export or dynamic import could have carried a new dependency past the check that confines `diff-match-patch` to one module.

## Known hazard the reviewer of a later phase must not inherit silently

**`detectPageContentFormat` only calls content HTML when the trimmed body both starts with `<` AND ends with `>`.** So a page whose markup gains an unclosed trailing element — routine in imported HTML, and reachable mid-edit — projects as raw text where the revision before it projected as stripped prose. Two incompatible coordinate systems for the same page, which would make `portAnchor` diff a stripped projection against raw markup and orphan or mislay every anchor on it.

I did **not** fix it here. The sniffer is shared with `diff-utils` and others, and the durable fix is a format stored per page rather than re-derived, which needs the schema a later phase adds. Changing the heuristic on a guess would move the failure rather than remove it. Instead it is pinned by a test, documented on `projectContent`, and written into `portAnchor`'s caller contract along with what `anchor.textHash` is for — a call site that cares can compare it against `hashText(oldText)` and detect the mismatch before trusting a port.

A related note for whoever wires this up: `portAnchor` receives two strings and cannot tell which page or revision they came from, so passing the wrong predecessor yields a confident wrong answer rather than an error. The hash is the guard against that, and applying it is the call site's job.

## Explicitly out of scope

No database changes, no migrations, no `packages/db`. No API routes, no UI. No tag vocabulary (`normalizeTagName`, `TAG_TARGETS`). Not wired into `applyPageMutation` or `convert-content-mode`.

That last one is a deliberate Phase-3 prerequisite, not an oversight, and it must land **before** any tag write path ships: because forward-porting is the *primary* mechanism, a missing save-path hook would silently degrade every edit to quote repair, which is the accuracy floor rather than the target. It is safe to omit now only because nothing writes anchors yet, so none can rot in the meantime.

Phase 3's bulk call site should also hoist the diff out of the per-anchor loop — `portAnchor` is a pure function of `(anchor, oldText, newText)` and recomputes it on every call, which is the right shape for the function and the wrong shape for a page carrying many anchors.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

https://claude.ai/code/session_017hRNwJ7LYfVn3g9W8sQRkv

### CODE REVIEW

File: packages/lib/src/content/anchoring/resolve.ts [hunk 1]
```diff
@@ -0,0 +1,184 @@
...
+ */
+
+import DiffMatchPatch from 'diff-match-patch';
+import type { AnchorResolution, TextAnchor } from './types';
+
+const dmp = new DiffMatchPatch();
+// Pinned rather than left to the library defaults so the fuzzy stage is
+// reproducible across dependency bumps.
+dmp.Match_Threshold = 0.5;
+dmp.Match_Distance = 1000;
```
[the remark this answers is not in the document]
Comment (2witstudios) [reply packages/lib/src/content/anchoring/resolve.ts]: Verified and fixed at all three sites — this was a real hole in the determinism the module doc claims. Fixed in 4960a4fe6.

I checked the library rather than taking the finding on trust. `diff_main` resolves its deadline as:

```js
if (this.Diff_Timeout <= 0) { opt_deadline = Number.MAX_VALUE; }
else { opt_deadline = (new Date).getTime() + this.Diff_Timeout * 1000; }
```

So the default of `1.0` does not fail on expiry — it returns a **valid but suboptimal** diff. That is the worst shape for this code, because nothing downstream can tell the difference.

**`resolve.ts`** — `dmp.Diff_Timeout = 0` alongside the other pinned knobs. `textSimilarity` could otherwise score the same pair of strings differently on a fast machine than on a loaded one, and a confidence wobbling across `FUZZY_SIMILARITY_FLOOR` would orphan a tag on one host while keeping it on another. Work stays bounded by the caller: it only ever compares a quote against a slice of its own length.

**`reanchor.ts`** — now `{ format: 'text', timeout: 0 }`. Worth being precise about why this one matters, since it isn't the obvious reason: the *ported offsets* are correct under any valid edit script, so a timed-out diff does not misplace the anchor. But `survived` counts characters per unchanged span, so a different **segmentation** changes the reported confidence and therefore the orphan decision at the floor. That is what was moving with the clock. (`timeout: 0` survives `diff-utils`' `options.timeout ?? 1000`, since `??` only defaults on nullish.)

**`reanchor.test.ts`** — kept the exact `confidence: 0.6` and recorded that it holds only because the deadline is pinned.

Two things beyond the literal finding:

1. A contract test now asserts both pins, and also that `diff-match-patch` is imported by `resolve.ts` **only**. The underlying risk isn't these two call sites — it's the next module that constructs its own `DiffMatchPatch` and silently inherits the defaults again. Mutating either pin turns that test red; I checked both go red rather than assuming.
2. While in here I also documented a cost the doc was missing: `portAnchor` diffs the whole document once per anchor. That's correct for a pure function of `(anchor, oldText, newText)` and wrong for a page carrying many anchors. The diff depends only on the two texts, so the bulk call site should hoist it — but that call site doesn't exist yet (wiring into `applyPageMutation` is a later phase), and inventing a batch API with no caller to shape it seemed worse than writing down where the hoist belongs.

Leaving this thread open for your verification rather than resolving it myself.

[the remark this answers is not in the document]
Comment (2witstudios) [reply packages/lib/src/content/anchoring/resolve.ts]: Correction to my previous reply on this thread, and a second change on top of it.

I justified `Diff_Timeout = 0` there with "the work stays bounded by the caller: `textSimilarity` only ever compares a quote against a slice of its own length." That was wrong. Nothing bounds a quote — `createAnchor(text, 0, text.length, …)` is a legal call for a user selecting a whole page — so removing the deadline replaced a determinism bug with an unbounded quadratic one.

I measured it against the vendored `diff-match-patch@1.0.5`, two fully dissimilar strings:

```
  1024 chars  unbounded      36ms   capped   36ms
  2048 chars  unbounded     142ms   capped  136ms
  4096 chars  unbounded     558ms   capped  545ms
  8192 chars  unbounded    2191ms   capped 1000ms
 20480 chars  unbounded   13887ms   capped 1001ms
```

`portAnchor` runs one whole-document diff **per anchor**, so a wholesale rewrite of a large page would have blocked the event loop for minutes.

Restoring the deadline is still not the fix, for the reason in my earlier reply: an expired deadline does not fail, it returns a differently *segmented* diff, so confidence would move with machine speed. So both bounds are now on the **input**, which is deterministic and bounded at the same time — fixed in 3f7cd202c:

- **`portAnchor`** strips the common prefix and suffix itself (linear) and degrades to the quote-and-context repair path when the remaining changed region exceeds 2048 characters. This costs nothing on a real save — the affixes strip everything but the keystrokes, and a 60KB document with a 14-character insertion measures **0ms**. The visible cost is that a large enough rewrite loses the duplicate-quote disambiguation, so there is now a test asserting the port and the search agree in that case, rather than leaving the trade implicit.
- **`textSimilarity`** is reached only for quotes up to 1024 characters (~36ms worst case); above that the fuzzy stage declines. Strategies 1 and 2 are linear and still run at any size, so a large intact quote that merely moved is still found.

The measurement behind the cap, `changedRegionSize`, is exported and tested directly — which edits keep the exact path is a policy worth being able to state rather than infer from a port's output.

Thanks for this thread; the original finding was correct and the follow-through turned up more than the finding itself.

File: packages/lib/src/content/anchoring/resolve.ts [hunk 2]
```diff
@@ -0,0 +1,264 @@
...
+  }
+  if (anchor.exact.length > 0) {
+    return true;
+  }
+  // A caret's own text is empty, so the slice check above passes at EVERY
+  // in-range offset — it is no evidence at all. Corroborate with the context,
+  // or a caret would claim maximum confidence against unrelated content.
+  const from = Math.max(0, anchor.start - anchor.prefix.length);
+  return text.slice(from, anchor.end + anchor.suffix.length) === anchor.prefix + anchor.suffix;
+}
```
[the remark this answers is not in the document]
Comment (2witstudios) [reply packages/lib/src/content/anchoring/resolve.ts]: Verified and fixed in cc62a5fe8 — and this one corrects a call I made two commits ago, so thank you for catching it.

I had removed the `start > end` guard deliberately, reasoning it was provably redundant: `slice` yields `''` for an inverted range, only a caret can equal `''`, and a caret then has to satisfy the context check. I even mutation-tested it, saw nothing go red, and took that as confirmation. What I missed is the case where the context is empty *too* — then the corroboration is vacuous as well, and your example passes every check:

```
{ start: 1, end: 0, exact: '', prefix: '', suffix: '' }  →  status 'exact', start 1, end 0
```

Restored, with a test that fails without it rather than a mutation that survived.

Adopted the integer check as well. The failure mode is worth stating because it isn't just hygiene: `slice` truncates rather than rejecting, so `end: END + 0.5` **matched the quote** and handed a fractional `end` straight back as a position for the caller to slice with. I test each side independently — with only one offset broken the remaining bounds checks all pass, so nothing else catches it, and a combined test would have let either half rot.

I kept the guards as an explicit list rather than folding them together, since the comment above them now has to explain four different ways `String.slice` forgives bad input. Anchors reach this function from a jsonb column, so untrusted offsets are the right default here rather than defensive noise.

Leaving this open for your verification.

File: packages/lib/src/content/anchoring/text-projection.ts [hunk 3]
```diff
@@ -0,0 +1,395 @@
...
+ * document, which would allocate a full copy of the page per element.
+ */
+function skipRawText(html: string, index: number, name: string): number {
+  for (let i = index; i < html.length; i += 1) {
+    if (html[i] !== '<' || html[i + 1] !== '/') {
+      continue;
+    }
+    if (html.slice(i + 2, i + 2 + name.length).toLowerCase() === name) {
+      return readTag(html, i).next;
+    }
```
[the remark this answers is not in the document]
Comment (2witstudios) [reply packages/lib/src/content/anchoring/text-projection.ts]: Confirmed and fixed in cc62a5fe8.

Reproduced before changing anything: `<p>a</p><script>x</scripture>leaked</script><p>b</p>` projected to `a\nleaked\nb` — the skip ended at `</scripture>`, and everything after it was projected as visible prose. Worse than cosmetic for this module, since the leaked text lands in the coordinate system anchors are measured against, so offsets after it are wrong too.

The closing name now has to end where it ends — the next character must be absent or not a tag-name character. Regression tests cover both `</scripture>` before the real `</script>` and `</styles>` before `</style>`, and mutating the boundary condition turns them red.

This was on my own list of things I'd noticed and waved off as too marginal to bother with, which was the wrong call — it costs one comparison. Leaving this open for your verification.

**Comment by 2witstudios** [comment]
@coderabbitai full review

The check status reads "Review rate limited", so I'd like a complete pass over the head commit rather than assume the incremental one was exhaustive. Several rounds of changes have landed since the first review — input bounds on both diff paths, offset validation in `positionHolds`, and four separate fixes in the HTML tokenizer.

Areas where I'd particularly value a second opinion:

1. `MAX_EXACT_DIFF_REGION = 2048` and `MAX_FUZZY_QUOTE = 1024` — both are bounds on input size chosen so the work stays deterministic without a wall-clock deadline. Are they in the right place, and is degrading to the quote-repair path above the threshold the right failure mode?
2. The hand-rolled HTML tokenizer in `text-projection.ts`. It exists because the projection must be byte-identical on server and client, so `DOMParser` is not an option. Divergence from what a browser actually renders is a correctness bug here, since offsets are measured against this output.
3. `positionHolds` — anchors arrive from a jsonb column, so every offset is untrusted input.

File: packages/lib/src/content/anchoring/__tests__/purity.test.ts [hunk 4]
```diff
@@ -0,0 +1,80 @@
...
+  it.each(MODULES)('%s imports nothing beyond the anchoring core and diff-match-patch', (name) => {
+    const src = readModule(name);
+    const specifiers = [...src.matchAll(/^import[^'"]*from ['"]([^'"]+)['"];?$/gm)].map((m) => m[1]);
+
+    for (const specifier of specifiers) {
+      if (specifier.startsWith('.')) {
+        continue;
+      }
+      expect(ALLOWED_PACKAGE_IMPORTS.has(specifier), `unexpected import: ${specifier}`).toBe(true);
+    }
```
[the remark this answers is not in the document]
Comment (2witstudios) [reply packages/lib/src/content/anchoring/__tests__/purity.test.ts]: Adopted in 04ec6d6de. `moduleSpecifiers` now covers `import … from`, bare side-effect `import`, `export … from`, `export * from`, dynamic `import()` and `require()`, and it has a test of its own asserting all seven forms are seen — the helper is now load-bearing for two other assertions, so it needed to stop being untested itself.

You're right about the shape of the risk. The check that confines `diff-match-patch` to a single module is worth more than the file-by-file purity check, because the real hazard isn't this PR — it's the next module that constructs its own `DiffMatchPatch` and silently inherits the one-second deadline that this whole thread series has been about. A side-effect or dynamic import sliding past that check is exactly how that arrives quietly.

I stopped short of a syntax-aware parser. Adding a TS AST dependency to a test that exists to prove this directory has *no* dependencies felt like the wrong trade, and the regex set now covers every form the module system offers. If it ever needs to survive obfuscation rather than ordinary drift, a parser is the right answer — but drift is the threat model here.

File: packages/lib/src/content/anchoring/reanchor.ts [hunk 5]
```diff
@@ -0,0 +1,256 @@
...
+    if (mapped === null) {
+      return orphaned();
+    }
+    return {
+      status: mapped === anchor.start ? 'exact' : 'shifted',
+      start: mapped,
+      end: mapped,
+      confidence: located.confidence,
+    };
+  }
```
[the remark this answers is not in the document]
Comment (2witstudios) [reply packages/lib/src/content/anchoring/reanchor.ts]: Correct, and fixed in 04ec6d6de.

The inconsistency was introduced by an earlier fix in this same review round: the caret branch predates carets being repairable at all. When `resolveAnchor` could only ever return `exact`/confidence 1 for a caret, `mapped === anchor.start` was sufficient. Once I taught it to repair carets from `prefix + suffix` — because a caret was claiming confidence 1 against unrelated content — `located.confidence` for a caret could be 0.95, and the branch had not been revisited.

Both branches now apply the same rule, and the caret counterpart of the range test is there: hint 2 too far right so it must be repaired, then an edit that shifts it back onto the very offset it recorded. Mutating the `located.confidence === 1` term turns it red.

The point about `CONFIDENCE.position` is the one that made this worth fixing rather than tidying: `'exact'` carrying confidence 0.95 from one code path and 1 from another means a consumer cannot gate on either field alone.

File: packages/lib/src/content/anchoring/resolve.ts [hunk 6]
```diff
@@ -0,0 +1,273 @@
...
+dmp.Match_Threshold = 0.5;
+dmp.Match_Distance = 1000;
+// 0 disables the deadline, which is the whole point: the library default of one
+// second makes diff_main return a SUBOPTIMAL diff when the clock runs out, so
+// the same two strings score differently on a fast machine than on a loaded
+// one. Anchors have to resolve identically everywhere, and a confidence that
+// wobbles across the FUZZY_SIMILARITY_FLOOR would orphan a tag on one host and
+// keep it on another. The work is bounded by the caller: textSimilarity only
+// ever compares a quote against a slice of its own length.
+dmp.Diff_Timeout = 0;
```
[the remark this answers is not in the document]
Comment (2witstudios) [reply packages/lib/src/content/anchoring/resolve.ts]: Agreed and fixed in 04ec6d6de — the bound now lives inside `textSimilarity`.

Your framing is the part I had wrong. I had put the cap at the call site in `resolveAnchor` and reasoned about `resolveAnchor`'s inputs, which is a bound only *some* callers observe, and `./content/anchoring/resolve` is in the package exports map so the others are reachable. That is not a bound.

`textSimilarity` now caps its own input at `MAX_SIMILARITY_INPUT` (1024) and returns 0 above it. Reporting 0 rather than throwing or approximating is deliberate: too large to compare is reported as dissimilar, which is the safe direction — an anchor orphans rather than claiming a match nothing verified.

That also let me delete the call-site check rather than keep both. 0 falls below `FUZZY_SIMILARITY_FLOOR`, so the fuzzy stage orphans by the same route it already used, and there is one threshold instead of two that have to be kept in step. Tests cover the oversized case from outside as well as through `resolveAnchor`, and mutating the cap turns them red.

The `Diff_Timeout` comment now names `MAX_SIMILARITY_INPUT` as what bounds the work, as you suggested.

File: packages/lib/src/content/anchoring/text-projection.ts [hunk 7]
```diff
@@ -0,0 +1,402 @@
...
+    // column. Leaving the entity as literal text keeps the projection storable.
+    const isSurrogate = code >= 0xd800 && code <= 0xdfff;
+    if (!Number.isFinite(code) || code < 1 || code > 0x10ffff || isSurrogate) {
+      return null;
+    }
+    return { value: String.fromCodePoint(code), length };
+  }
+
+  const named = NAMED_ENTITIES[body.toLowerCase()];
+  return named === undefined ? null : { value: named, length };
```
[the remark this answers is not in the document]
Comment (2witstudios) [reply packages/lib/src/content/anchoring/text-projection.ts]: Confirmed and fixed in 04ec6d6de — this was the most serious of the four. One correction on the symptom, because I reproduced it before changing anything and it is worse than "visible garbage".

Only `&__proto__;` actually gets through. `valueOf` and `toString` are saved by `body.toLowerCase()` — `Object.prototype` has no `valueof` or `tostring` — and `constructor` and `hasOwnProperty` exceed the entity length guard. So:

```
&__proto__;       "a b"                  ← decoded
&valueOf;         "a&valueOf;b"          ← safe (lowercasing)
&hasOwnProperty;  "a&hasOwnProperty;b"   ← safe (length guard)
```

And it does not project as `[object Object]`. `Object.prototype` was accepted as a decoded value, then handed to `isWhitespace`, whose regex was **unanchored** — so it tested `[object Object]`, found the space inside it, and classified the entity as whitespace. The entity was silently **swallowed**, which is worse than garbage: garbage is visible, whereas this quietly puts every offset after it out by eleven characters against a projection that still looks entirely reasonable.

Two fixes, since there were two independent faults:

- `NAMED_ENTITIES` is a `ReadonlyMap`, so the prototype is not part of the lookup at all. This is the real fix — and worth noting that `Readonly<Record<string, string>>` actively concealed it, since the compiler was happy to type a prototype method as `string`.
- The whitespace regex is anchored, so a predicate named for a single character can no longer match a substring of a longer one.

Regression tests cover `&__proto__;`, `&valueOf;` and `&hasOwnProperty;`, plus the real entities still decoding. In fairness the second fix is defence in depth: with the Map in place I could not construct an input that reaches `isWhitespace` with more than one character, so I have not been able to make a test fail for the anchoring alone, and I am not claiming it as verified.
